tiledb/api/src/column.rs

```diff
@@ -0,0 +1,509 @@
+            )
+            .map_err(|e| {
+                Error::from(format!(
+                    "Invalid UTF-8 in column data. \
+                    Use to_boxed_u8_vec instead: {}",
+                    e
+                ))
+            })?;
+
+            ret.push(val);
+        }
+
+        Ok(ret)
+    }
+
+    pub fn to_boxed_u8_vec(
+        &self,
+        datatype: Datatype,
+        cell_val_size: u32,
+    ) -> TileDBResult<Vec<Box<[u8]>>> {
+        // Currently we're only supporting ASCII and UTF-8 strings.
+        if !matches!(datatype, Datatype::StringAscii | Datatype::StringUtf8) {
+            return Err(Error::from(format!(
+                "Invalid datatype '{}' for to_string_vec",
+                datatype.to_string().unwrap_or("<unknown type>".to_owned())
+            )));
+        }
+
+        // For var sized data, use the to_type_vec methods for conversion.
+        if cell_val_size != u32::MAX {
+            return Err(Error::from("Use to_vec<T> for fixed sized data."));
+        }
+
+        if self.offsets().is_none() {
+            return Ok(Vec::new());
+        }
+
+        let mut ret: Vec<Box<[u8]>> = Vec::new();
+
+        let data = self.data();
+        let offsets = self.offsets().unwrap();
+        let num_offsets = offsets.len();
+        for (idx, offset) in offsets.iter().enumerate() {
+            let len = if idx == num_offsets - 1 {
+                self.data().len() as u64 - *offset
+            } else {
+                offset - offsets[idx - 1]
+            };
+
+            let val = data[(*offset as usize)..((*offset + len) as usize)]
+                .to_vec()
+                .into_boxed_slice();
+
+            ret.push(val);
+        }
+
+        Ok(ret)
+    }
+
+    pub fn num_values(&self) -> u64 {
+        match self {
+            Column::Referenced(col) => col.num_values,
+            Column::Allocated(col) => col.num_values,
+        }
+    }
+
+    pub fn value_size(&self) -> u64 {
+        match self {
+            Column::Referenced(col) => col.value_size,
+            Column::Allocated(col) => col.value_size,
+        }
+    }
+
+    pub fn data(&self) -> &[u8] {
+        match self {
+            Column::Referenced(col) => col.data,
+            Column::Allocated(col) => &col.data[..],
+        }
+    }
+
+    pub fn offsets(&self) -> Option<&[u64]> {
+        match self {
+            Column::Referenced(col) => col.offsets,
+            Column::Allocated(col) => match &col.offsets {
+                Some(offsets) => unsafe {
+                    Some(std::slice::from_raw_parts(
+                        offsets.as_ptr(),
+                        self.num_values() as usize,
+                    ))
+                },
+                None => None,
+            },
+        }
+    }
+
+    pub fn is_allocated(&self) -> bool {
+        match self {
+            Column::Referenced(_) => false,
+            Column::Allocated(_) => true,
+        }
+    }
+
+    pub fn to_json(
```

Comment:
This is probably better done as a `impl Serialize` right?  Then the JSON serializer will pick it up.

Reply:
Originally I had this bit of logic in the Enumeration Debug trait and realized that it'd make a lot more sense have a Debug for Column trait. And then I ran smack dab into the issue I noted above about Column's not knowing their datatype or cell value number so I had to make it a callable method to use in the Enumeration Debug trait.